File: scripts/generate-scenes-data.ts
/**
 * generate-scenes-data.ts
 *
 * 読書シーン別の作品JSONを生成する。
 *
 * 入力:
 *   public/data/works-list.json    (WorkListItem[])
 *   data/normalized/works.json     (Work[] — discoveryAttributes参照用)
 *   public/data/work-id-map.json   (workId → fileId マッピング)
 *   src/constants/readingScenes.ts (シーン定義)
 *
 * 出力:
 *   public/data/scenes/index.json          (SceneIndexMeta)
 *   public/data/scenes/{slug}.json         (SceneWorksData)
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { join } from "path";
import { READING_SCENES, type ReadingScene } from "../src/constants/readingScenes.js";
import type { WorkListItem, SceneWorksData, SceneIndexMeta } from "../src/types/work.js";

// ── パス設定 ──────────────────────────────────────────────────────
const ROOT = process.cwd();
const SCENES_DIR = join(ROOT, "public", "data", "scenes");
mkdirSync(SCENES_DIR, { recursive: true });

// ── データ読み込み ────────────────────────────────────────────────
const worksList: WorkListItem[] = JSON.parse(
  readFileSync(join(ROOT, "public", "data", "works-list.json"), "utf-8")
);

// normalized works — discoveryAttributes (paceTag, depthTag) 取得用
const normalizedWorks: Array<{
  workId: string;
  discoveryAttributes: Record<string, unknown>;
  volumeCount: number;
  type: string;
}> = JSON.parse(
  readFileSync(join(ROOT, "data", "normalized", "works.json"), "utf-8")
);

// workId(normalized) → fileId マップ
const workIdMap: Record<string, string> = JSON.parse(
  readFileSync(join(ROOT, "public", "data", "work-id-map.json"), "utf-8")
);

// fileId → discoveryAttributes マップを構築
const fileIdToAttrs = new Map<string, Record<string, unknown>>();
for (const w of normalizedWorks) {
  const fileId = workIdMap[w.workId];
  if (fileId) {
    fileIdToAttrs.set(fileId, w.discoveryAttributes ?? {});
  }
}

console.log(`読み込み: ${worksList.length} 作品`);
console.log(`属性マップ: ${fileIdToAttrs.size} エントリ`);

// ── スコアリング ──────────────────────────────────────────────────

/**
 * 1作品の読書シーン適合スコアを算出する。
 *
 * スコア内訳:
 *   primaryTags マッチ    +3 per tag
 *   bonusTags マッチ      +1 per tag
 *   paceTag マッチ        +1
 *   depthTag マッチ       +1
 *   volumeCount 条件      +1
 *   preferredType 一致    +1
 *   excludeTags マッチ    → 即 -999（除外）
 *
 * タグが0件の作品は構造的ヒューリスティックのみで判定する。
 */
function scoreWork(work: WorkListItem, scene: ReadingScene): number {
  const tags = new Set(work.discoveryTags);
  const attrs = fileIdToAttrs.get(work.workId) ?? {};

  // 除外タグチェック
  for (const ex of scene.excludeTags) {
    if (tags.has(ex)) return -999;
  }

  let score = 0;

  // primaryTags
  for (const t of scene.primaryTags) {
    if (tags.has(t)) score += 3;
  }

  // bonusTags
  for (const t of scene.bonusTags) {
    if (tags.has(t)) score += 1;
  }

  // paceTag
  if (scene.paceTags && attrs.paceTag) {
    if (scene.paceTags.includes(attrs.paceTag as string)) score += 1;
  }

  // depthTag
  if (scene.depthTags && attrs.depthTag) {
    if (scene.depthTags.includes(attrs.depthTag as string)) score += 1;
  }

  // volumeCount 条件
  if (scene.volumeCountMin !== undefined && work.volumeCount >= scene.volumeCountMin) score += 1;
  if (scene.volumeCountMax !== undefined && work.volumeCount <= scene.volumeCountMax) score += 1;

  // preferredType
  if (scene.preferredType && work.type === scene.preferredType) score += 1;

  // ── タグなし作品のフォールバック ──────────────────────────────
  // discoveryTags が空の場合、構造的ヒューリスティックのみで基礎スコアを付与
  if (tags.size === 0) {
    score += structuralScore(work, scene);
  }

  return score;
}

/**
 * タグ未設定作品向け: 作品の構造情報からシーン適合度を推定する。
 * 0〜2 点を返す（主タグマッチの最低基準=3点には届かないが、他の条件と合わせて合格できる）。
 */
function structuralScore(work: WorkListItem, scene: ReadingScene): number {
  let s = 0;
  const { slug, volumeCount, type } = { ...work, slug: "" };
  const sc = scene.slug;

  // 巻数ベース
  if (sc === "commute" && work.volumeCount <= 10) s += 1;
  if (sc === "short-break" && work.volumeCount <= 3) s += 1;
  if (sc === "holiday-binge" && work.volumeCount >= 8) s += 1;

  // タイプベース
  if ((sc === "before-sleep" || sc === "cafe" || sc === "think-deeply") && work.type === "novel") s += 1;
  if ((sc === "stress-relief" || sc === "exciting") && work.type === "manga") s += 1;

  // 完結ボーナス（一気読み・すきま向き）
  if ((sc === "short-break" || sc === "commute") && work.status === "completed") s += 1;

  // 連載中ボーナス（ワクワク・一気読み向き）
  if ((sc === "exciting" || sc === "holiday-binge") && work.status === "ongoing") s += 1;

  return s;
}

// ── シーンごとに作品選定・JSON生成 ───────────────────────────────

/** 1シーンあたりの最大作品数 */
const MAX_WORKS_PER_SCENE = 120;
/** タグ/属性マッチで必要な最低スコア（これ未満はタグなし構造マッチで補完） */
const PRIMARY_MIN_SCORE = 3;
/** 構造マッチだけで補完する際の最低スコア */
const STRUCTURAL_MIN_SCORE = 1;
/** 構造マッチで補完を開始する閾値（タグマッチがこの件数未満なら補完する） */
const FALLBACK_THRESHOLD = 30;

function sortWorks(a: { score: number; work: WorkListItem }, b: { score: number; work: WorkListItem }) {
  if (b.score !== a.score) return b.score - a.score;
  const dateA = a.work.latestPublishedDate ?? "";
  const dateB = b.work.latestPublishedDate ?? "";
  if (dateB !== dateA) return dateB.localeCompare(dateA);
  return a.work.title.localeCompare(b.work.title, "ja");
}

const sceneIndexItems: SceneIndexMeta["scenes"] = [];

for (const scene of READING_SCENES) {
  // Step1: タグ/属性マッチ（スコアが primaryTags 等から来るもの）
  const allScored = worksList.map((work) => ({ work, score: scoreWork(work, scene) }));
  const excluded = allScored.filter(({ score }) => score <= -999);
  const excludedIds = new Set(excluded.map(({ work }) => work.workId));

  // タグ or 属性マッチしたもの（discoveryTags が空でない OR attrs マッチ）
  const tagged = allScored.filter(({ work, score }) =>
    score >= PRIMARY_MIN_SCORE && !excludedIds.has(work.workId)
  ).sort(sortWorks);

  let combined = tagged;

  // タグマッチが少ない場合は構造ベースで補完
  if (tagged.length < FALLBACK_THRESHOLD) {
    const taggedIds = new Set(tagged.map(({ work }) => work.workId));
    const structural = allScored.filter(({ work, score }) =>
      score >= STRUCTURAL_MIN_SCORE &&
      score < PRIMARY_MIN_SCORE &&
      !excludedIds.has(work.workId) &&
      !taggedIds.has(work.workId)
    ).sort(sortWorks);
    combined = [...tagged, ...structural];
  }

  const items = combined.slice(0, MAX_WORKS_PER_SCENE).map(({ work }) => work);

  const data: SceneWorksData = {
    slug: scene.slug,
    label: scene.label,
    icon: scene.icon,
    description: scene.description,
    works: items,
    totalCount: items.length,
    generatedAt: new Date().toISOString(),
  };

  writeFileSync(
    join(SCENES_DIR, `${scene.slug}.json`),
    JSON.stringify(data)
  );

  sceneIndexItems.push({
    slug: scene.slug,
    label: scene.label,
    icon: scene.icon,
    description: scene.description,
    count: items.length,
  });

  console.log(`✓ scenes/${scene.slug}.json  (${items.length} 作品) [${scene.label}]`);
}

// ── index.json 生成 ───────────────────────────────────────────────

const indexData: SceneIndexMeta = {
  scenes: sceneIndexItems,
  generatedAt: new Date().toISOString(),
};

writeFileSync(
  join(SCENES_DIR, "index.json"),
  JSON.stringify(indexData)
);

console.log(`\n✓ scenes/index.json  (${READING_SCENES.length} シーン)`);
console.log("\nDone. 生成物 → public/data/scenes/");

File: scripts/generate-scenes-data.py
#!/usr/bin/env python3
"""
generate-scenes-data.ts の Python版。
読書シーン別の作品JSONを生成する。
"""

import json
import os
from datetime import datetime, timezone

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SCENES_DIR = os.path.join(ROOT, "public", "data", "scenes")
os.makedirs(SCENES_DIR, exist_ok=True)

# ── READING_SCENES 定義 (readingScenes.ts と同一) ──
READING_SCENES = [
    {
        "slug": "commute", "label": "通勤・通学", "icon": "🚃",
        "description": "電車・バスの移動中にサクッと読める",
        "primaryTags": ["読みやすい", "一気読み", "短編", "笑える", "爽快"],
        "bonusTags": ["明るい", "前向き", "日常系"],
        "excludeTags": ["深い", "絶望"],
        "paceTags": ["早い", "テンポ良い"], "depthTags": ["軽い"],
        "volumeCountMax": 12,
    },
    {
        "slug": "before-sleep", "label": "寝る前", "icon": "🌙",
        "description": "就寝前にほっこり・癒やされる作品",
        "primaryTags": ["癒やし", "穏やか", "優しい", "日常系", "心温まる"],
        "bonusTags": ["明るい", "切ない", "感動"],
        "excludeTags": ["怖い", "絶望", "ダーク", "バトル"],
        "depthTags": ["軽い", "中程度"], "preferredType": "novel",
    },
    {
        "slug": "holiday-binge", "label": "休日に一気読み", "icon": "📚",
        "description": "休みの日にどっぷり没入して読み切る",
        "primaryTags": ["一気読み", "世界観重視", "熱い", "バトル", "感動"],
        "bonusTags": ["考えさせられる", "泣ける", "深い"],
        "excludeTags": [], "volumeCountMin": 5,
    },
    {
        "slug": "short-break", "label": "すきま時間", "icon": "⏱️",
        "description": "5〜15分の合間にサッと読める",
        "primaryTags": ["短編", "読みやすい", "笑える", "日常系"],
        "bonusTags": ["明るい", "爽快", "癒やし"],
        "excludeTags": ["世界観重視", "深い", "絶望"],
        "paceTags": ["早い"], "depthTags": ["軽い"], "volumeCountMax": 5,
    },
    {
        "slug": "cafe", "label": "カフェでゆっくり", "icon": "☕",
        "description": "カフェで落ち着いてじっくり味わえる",
        "primaryTags": ["穏やか", "癒やし", "日常系", "心温まる", "優しい", "感動"],
        "bonusTags": ["切ない", "考えさせられる", "明るい"],
        "excludeTags": ["バトル", "絶望", "怖い", "熱い"],
        "depthTags": ["軽い", "中程度"], "preferredType": "novel",
    },
    {
        "slug": "stress-relief", "label": "ストレス解消したい", "icon": "💪",
        "description": "スカッと発散できる爽快・熱い作品",
        "primaryTags": ["笑える", "爽快", "熱い", "前向き", "やる気が出る", "バトル"],
        "bonusTags": ["明るい", "一気読み"],
        "excludeTags": ["ダーク", "絶望", "怖い"],
        "paceTags": ["早い"], "preferredType": "manga",
    },
    {
        "slug": "calm-down", "label": "気分を落ち着けたい", "icon": "🍃",
        "description": "心を穏やかにしてくれる癒やし系",
        "primaryTags": ["癒やし", "優しい", "穏やか", "心温まる", "日常系"],
        "bonusTags": ["感動", "泣ける", "切ない"],
        "excludeTags": ["バトル", "怖い", "絶望", "ダーク"],
        "depthTags": ["軽い", "中程度"],
    },
    {
        "slug": "exciting", "label": "ワクワクしたい", "icon": "✨",
        "description": "ドキドキ・ワクワクが止まらない作品",
        "primaryTags": ["熱い", "爽快", "バトル", "ファンタジー", "やる気が出る", "世界観重視", "一気読み"],
        "bonusTags": ["明るい", "前向き", "感動"],
        "excludeTags": ["絶望"],
        "paceTags": ["早い"], "preferredType": "manga",
    },
    {
        "slug": "think-deeply", "label": "考えたい", "icon": "🧠",
        "description": "読み終わった後も考え続けてしまう作品",
        "primaryTags": ["考えさせられる", "深い", "学べる", "世界観重視", "ダーク"],
        "bonusTags": ["感動", "泣ける", "絶望", "切ない"],
        "excludeTags": ["笑える"],
        "depthTags": ["重い", "中程度"], "preferredType": "novel",
    },
]

# ── データ読み込み ──
with open(os.path.join(ROOT, "public", "data", "works-list.json"), encoding="utf-8") as f:
    works_list = json.load(f)

with open(os.path.join(ROOT, "data", "normalized", "works.json"), encoding="utf-8") as f:
    normalized_works = json.load(f)

with open(os.path.join(ROOT, "public", "data", "work-id-map.json"), encoding="utf-8") as f:
    work_id_map = json.load(f)

# fileId → discoveryAttributes
file_id_to_attrs = {}
for w in normalized_works:
    fid = work_id_map.get(w["workId"])
    if fid:
        file_id_to_attrs[fid] = w.get("discoveryAttributes", {})

print(f"読み込み: {len(works_list)} 作品, 属性マップ: {len(file_id_to_attrs)} エントリ")


def score_work(work, scene):
    tags = set(work.get("discoveryTags", []))
    attrs = file_id_to_attrs.get(work["workId"], {})

    for ex in scene.get("excludeTags", []):
        if ex in tags:
            return -999

    score = 0
    for t in scene.get("primaryTags", []):
        if t in tags:
            score += 3
    for t in scene.get("bonusTags", []):
        if t in tags:
            score += 1

    if scene.get("paceTags") and attrs.get("paceTag"):
        if attrs["paceTag"] in scene["paceTags"]:
            score += 1
    if scene.get("depthTags") and attrs.get("depthTag"):
        if attrs["depthTag"] in scene["depthTags"]:
            score += 1

    if scene.get("volumeCountMin") is not None and work.get("volumeCount", 1) >= scene["volumeCountMin"]:
        score += 1
    if scene.get("volumeCountMax") is not None and work.get("volumeCount", 1) <= scene["volumeCountMax"]:
        score += 1

    if scene.get("preferredType") and work.get("type") == scene["preferredType"]:
        score += 1

    if len(tags) == 0:
        score += structural_score(work, scene)

    return score


def structural_score(work, scene):
    s = 0
    sc = scene["slug"]
    vc = work.get("volumeCount", 1)
    wtype = work.get("type", "")
    status = work.get("status", "")

    if sc == "commute" and vc <= 10: s += 1
    if sc == "short-break" and vc <= 3: s += 1
    if sc == "holiday-binge" and vc >= 8: s += 1

    if sc in ("before-sleep", "cafe", "think-deeply") and wtype == "novel": s += 1
    if sc in ("stress-relief", "exciting") and wtype == "manga": s += 1

    if sc in ("short-break", "commute") and status == "completed": s += 1
    if sc in ("exciting", "holiday-binge") and status == "ongoing": s += 1

    return s


MAX_WORKS = 120
PRIMARY_MIN = 3
STRUCTURAL_MIN = 1
FALLBACK_THRESHOLD = 30

scene_index_items = []

for scene in READING_SCENES:
    scored = [(w, score_work(w, scene)) for w in works_list]
    excluded_ids = {w["workId"] for w, s in scored if s <= -999}

    def sort_key(x):
        return (-x[1], x[0].get("latestPublishedDate") or "", x[0]["title"])

    tagged = sorted(
        [(w, s) for w, s in scored if s >= PRIMARY_MIN and w["workId"] not in excluded_ids],
        key=sort_key, reverse=False
    )
    # Re-sort: score desc, date desc, title asc
    tagged.sort(key=lambda x: (-x[1], -(ord(x[0].get("latestPublishedDate", "0")[0]) if x[0].get("latestPublishedDate") else 0)))

    combined = tagged
    if len(tagged) < FALLBACK_THRESHOLD:
        tagged_ids = {w["workId"] for w, _ in tagged}
        structural = sorted(
            [(w, s) for w, s in scored if STRUCTURAL_MIN <= s < PRIMARY_MIN
             and w["workId"] not in excluded_ids and w["workId"] not in tagged_ids],
            key=lambda x: (-x[1], x[0].get("title", ""))
        )
        combined = tagged + structural

    items = [w for w, _ in combined[:MAX_WORKS]]

    data = {
        "slug": scene["slug"], "label": scene["label"], "icon": scene["icon"],
        "description": scene["description"], "works": items,
        "totalCount": len(items),
        "generatedAt": datetime.now(timezone.utc).isoformat(),
    }

    with open(os.path.join(SCENES_DIR, f'{scene["slug"]}.json'), "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, separators=(",", ":"))

    scene_index_items.append({
        "slug": scene["slug"], "label": scene["label"], "icon": scene["icon"],
        "description": scene["description"], "count": len(items),
    })

    print(f'✓ scenes/{scene["slug"]}.json  ({len(items)} 作品) [{scene["label"]}]')

index_data = {
    "scenes": scene_index_items,
    "generatedAt": datetime.now(timezone.utc).isoformat(),
}
with open(os.path.join(SCENES_DIR, "index.json"), "w", encoding="utf-8") as f:
    json.dump(index_data, f, ensure_ascii=False, separators=(",", ":"))

print(f"\n✓ scenes/index.json  ({len(READING_SCENES)} シーン)")
